### PR TITLE
Limit the length of rd.zfcp when refresh bootmap

### PR DESCRIPTION
**Pull request by changzhi1990**
In the boot from volume scenario for RHEL, the VM may fail if the length of kernel parameter line exceeds the maximum allowed. So the production recommend is 8 different paths of rd.zfcp.

If the count of FCP devices is greater than 8, it will report an error and exit directly.

If the count of WWPN is greater than 8, it will report a warning and don't exit.

### CODE REVIEW

**Comment by changzhi1990** [comment]
Boot from volume maybe fails at the refreshbootmap stage for length of kernel parameter line exceed the maximum allowed. So the production recommend is 8 paths of rd.zfcp.

If the count of FCP devices is greater than 8, it will report an error and exit directly.

If the count of WWPN is greater than 8, it will report a warning and don't exit.

**Comment by changzhi1990** [comment]
With this pr, if the path is less than 8, it would work fine. If the path is greater than 8, the output is:
[root@rhel82-com ~]# sudo /opt/zthin/bin/refresh_bootmap --fcpchannel=5c73,5d73,5c74,5c71,5d74 --wwpn=5005076802100C1A,5005076802200C1A,5005076802400C1A,5005076802300C1A,5005076802300C1B,5005076802200C1B,5005076802100C1B,5005076802400C1B --lun=0000000000000000
zthinUserID is CHAZ00D4
Error: non-zero CP response for command 'q fcp': #40
Done
Done
Done
Done
Done
These WWPNs will be operated: 5005076802400c1a 5005076802400c1b
/dev/disk/by-path/ccw-0.0.5c73-zfcp-0x5005076802400c1a:0x0000000000000000 is: /dev/disk/by-path/ccw-0.0.5c73-zfcp-0x5005076802400c1a:0x0000000000000000
/dev/disk/by-path/ccw-0.0.5d73-zfcp-0x5005076802400c1a:0x0000000000000000 is: /dev/disk/by-path/ccw-0.0.5d73-zfcp-0x5005076802400c1a:0x0000000000000000
/dev/disk/by-path/ccw-0.0.5c74-zfcp-0x5005076802400c1a:0x0000000000000000 is: /dev/disk/by-path/ccw-0.0.5c74-zfcp-0x5005076802400c1a:0x0000000000000000
/dev/disk/by-path/ccw-0.0.5d74-zfcp-0x5005076802400c1a:0x0000000000000000 is: /dev/disk/by-path/ccw-0.0.5d74-zfcp-0x5005076802400c1a:0x0000000000000000
Multipath is not enabled, using single path devNode: /dev/disk/by-path/ccw-0.0.5d74-zfcp-0x5005076802400c1a:0x0000000000000000-part1.
Begin to refreshZIPL.
devNode is: /dev/disk/by-path/ccw-0.0.5d74-zfcp-0x5005076802400c1a:0x0000000000000000-part1 point: /mnt/sh5dH
The count of FCP devices greater than WWPN.
Use as much as possible of FCP and WWPN rd.zfcp: rd.zfcp=0.0.5d74,0x5005076802400c1a,0x0000000000000000 rd.zfcp=0.0.5d73,0x5005076802400c1b,0x0000000000000000 rd.zfcp=0.0.5c74,0x5005076802400c1b,0x0000000000000000 rd.zfcp=0.0.5c71,0x5005076802400c1a,0x0000000000000000 rd.zfcp=0.0.5c73,0x5005076802400c1b,0x0000000000000000 
Refresh rhel78 zipl.
Updating /mnt/sh5dH/etc/fstab
Move stale multipath bindings and wwids if any
WWPNs: 5005076802400c1a 5005076802400c1b
FCPs: 5d74 5d73 5c74 5c71 5c73
Begin to cleanup umount dir.
Finish cleanuping umount dir.
Begin to disconnect FCPs
Disconnect FCP device: 5d74
Disconnect FCP device: 5d74
FCP device 5d74 has already detached.
Disconnect FCP device: 5d73
Disconnect FCP device: 5d73
FCP device 5d73 has already detached.
Disconnect FCP device: 5c74
Disconnect FCP device: 5c74
FCP device 5c74 has already detached.
Disconnect FCP device: 5c71
Disconnect FCP device: 5c71
FCP device 5c71 has already detached.
Disconnect FCP device: 5c73
Disconnect FCP device: 5c73
FCP device 5c73 has already detached.
Finish disconnecting FCPs.
Begin to refresh multipath bindings.
Finish refreshing multipath bindings.

File: zthin-parts/zthin/bin/refresh_bootmap [hunk 1]
```diff
@@ -295,14 +295,59 @@ function refreshZIPL {
...
   # Prepare the new rd.zfcp parameter value
+  fcps_len=${#fcps[@]}
+  wwpns_len=${#wwpns[@]}
+  res=$(( $fcps_len * $wwpns_len ))
   rd_zfcp=""
-  for fcp in ${fcps[@]}
-  do
-    for w in ${wwpns[@]}
+
+  # For rhel8.x, the max support items of "rd.zfcp" is 8, so we need to judge the
```
Comment (dongyanyang) [inline zthin-parts/zthin/bin/refresh_bootmap:303]: not only rhel8.x

File: zthin-parts/zthin/bin/refresh_bootmap [hunk 2]
```diff
@@ -295,14 +295,59 @@ function refreshZIPL {
     os=$os$version
   fi
   # Prepare the new rd.zfcp parameter value
+  fcps_len=${#fcps[@]}
+  wwpns_len=${#wwpns[@]}
+  res=$(( $fcps_len * $wwpns_len ))
```
Comment (dongyanyang) [inline zthin-parts/zthin/bin/refresh_bootmap:300]: I assume we have handled the case when fcps_len or wwpns_len is zero?

File: zthin-parts/zthin/bin/refresh_bootmap [hunk 3]
```diff
@@ -411,6 +456,21 @@ function refreshFCPBootmap {
...
+  # Check the count of FCP devices, if greater than 8, report error and exit.
+  fcps_len=${#fcps[@]}
+  if [ $fcps_len -gt 8 ]
+  then
+      inform "ERROR: The count of FCP devices can't greater than 8."
+      exit
+  fi
+
+  # Check the count of WWPN, if greater than 8, warning and don't exit.
+  wwpns_len=${#wwpns[@]}
```
Comment (dongyanyang) [inline zthin-parts/zthin/bin/refresh_bootmap:468]: should check after get the real wwpn, here the wwpn is the passed in parameters.

File: zthin-parts/zthin/bin/refresh_bootmap [hunk 4]
```diff
@@ -411,6 +456,21 @@ function refreshFCPBootmap {
   IFS=',' read -r -a fcps <<< "$fcpchannels"
   IFS=',' read -r -a ws <<< "$wwpn"
 
+  # Check the count of FCP devices, if greater than 8, report error and exit.
+  fcps_len=${#fcps[@]}
+  if [ $fcps_len -gt 8 ]
+  then
+      inform "ERROR: The count of FCP devices can't greater than 8."
```
Comment (dongyanyang) [inline zthin-parts/zthin/bin/refresh_bootmap:463]: The FCP channel count %d is greater than 8.

File: zthin-parts/zthin/bin/refresh_bootmap [hunk 5]
```diff
@@ -295,14 +295,59 @@ function refreshZIPL {
...
+      do
+        rd_zfcp+="rd.zfcp=0.0."$fcp,0x$w,$lun" "
+      done
     done
-  done
+    inform "Do a fullmesh of FCP devices and WWPNs for rd.zfcp: $rd_zfcp"
+  # If the result of fcps_len multiply wwpns_len greater than 8, we need to use
+  # every FCP device and as much as possible to use WWPN.
+  else
+    if [ $fcps_len -gt $wwpns_len ];then
```
Comment (dongyanyang) [inline zthin-parts/zthin/bin/refresh_bootmap:321]: These code logic can be simplified.

File: zthin-parts/zthin/bin/refresh_bootmap (hunk 2 above)
Comment (changzhi1990) [reply zthin-parts/zthin/bin/refresh_bootmap:300]: yes

**Comment by dongyanyang** [comment]
@bjhuangr @jichenjc pls help to review, thanks

**Comment by bjhuangr** [comment]
@changzhi1990 The code change looks good to me. Please add some words to describe the change into commit message.

File: zthin-parts/zthin/bin/refresh_bootmap [hunk 6]
```diff
@@ -295,14 +295,58 @@ function refreshZIPL {
...
+    do
+        if [ $fcp_index -ge $fcps_len ]; then
+          fcp_index=0
+        fi
+
+        if [ $wwpn_index -ge $wwpns_len ]; then
+          wwpn_index=0
+        fi
+
+        x=${fcps[$fcp_index]},0x${wwpns[$wwpn_index]}" "
```
Comment (dongyanyang) [inline zthin-parts/zthin/bin/refresh_bootmap:334]: missing $lun?

**Comment by changzhi1990** [comment]
For 4 FCPs and 4 WWPNs scenarios, the output is:

Aug 10 02:33:19 112-cmp-cz root[31258]: 28492.0 refresh_bootmap: refresh_bootmap: refreshZIPL: Use as much as possible of FCP devices: 1d0c 1b0b 1c09 1a09 and WWPN: 500507680b25bac6 500507680b25bac7 500507680b26bac6 500507680b26bac7 rd.zfcp: rd.zfcp=0.0.1d0c,0x500507680b25bac6,0x0000000000000000 rd.zfcp=0.0.1b0b,0x500507680b25bac7,0x0000000000000000 rd.zfcp=0.0.1c09,0x500507680b26bac6,0x0000000000000000 rd.zfcp=0.0.1a09,0x500507680b26bac7,0x0000000000000000 rd.zfcp=0.0.1d0c,0x500507680b25bac7,0x0000000000000000 rd.zfcp=0.0.1b0b,0x500507680b26bac6,0x0000000000000000 rd.zfcp=0.0.1c09,0x500507680b26bac7,0x0000000000000000 rd.zfcp=0.0.1a09,0x500507680b25bac6,0x0000000000000000

The VM works fine:
image

**Comment by changzhi1990** [comment]
> Please add the commit message

done

**Pull request merged**